src/java.base/windows/native/libnio/ch/FileChannelImpl.c

```diff
@@ -145,6 +146,45 @@ Java_sun_nio_ch_FileChannelImpl_unmap0(JNIEnv *env, jobject this,
     return 0;
 }
 
+#define READ_WRITE_TRANSFER_SIZE  32768
+#define READ_WRITE_TRANSFER_LIMIT 2097152
+
+DWORD transfer_read_write(JNIEnv* env, HANDLE src, DWORD position, DWORD count,
+                          HANDLE dst)
+{
+    LARGE_INTEGER where;
+    where.QuadPart = (LONGLONG)position;
+    if (SetFilePointerEx(src, where, &where, FILE_BEGIN) == 0) {
+        JNU_ThrowIOExceptionWithLastError(env, "SetFilePointerEx failed");
+        return IOS_THROWN;
+    }
+
+    char buf[READ_WRITE_TRANSFER_SIZE];
+
+    DWORD tw = 0;
+    while (tw < count) {
+        DWORD remaining = count - tw;
+        DWORD nr = remaining < READ_WRITE_TRANSFER_SIZE ?
+            remaining : READ_WRITE_TRANSFER_SIZE;
+        if (ReadFile(src, (LPVOID)&buf, nr, &nr, NULL) == 0) {
+            JNU_ThrowIOExceptionWithLastError(env, "ReadFile failed");
+            return IOS_THROWN;
+        }
+
+        DWORD nw = 0;
+        if (WriteFile(dst, &buf, nr, &nw, NULL) == 0) {
+            JNU_ThrowIOExceptionWithLastError(env, "WriteFile failed");
+            return IOS_THROWN;
```

Comment:
If ReadFile or WriteFile fails then I assume we should return tw when > 0.

Reply:
In the macOS code if `read()` or `write()` fails then an exception is thrown. I suppose the two read-write versions should be consistent one way or the other.

Reply:
This is resolved in commit three.